File: WINGS_AI_Bootcamp/Elixir/collatz1.exs
defmodule Collatz do
  def next(k) do
    if rem(k, 2) == 0 do
      div(k, 2)
    else
      3 * k + 1
    end
  end

  def sequence(n) do
    generate_sequence(n, [n])
  end

  defp generate_sequence(1, seq), do: Enum.reverse(seq)

  defp generate_sequence(k, seq) do
    next_number = next(k)
    generate_sequence(next_number, [next_number | seq])
  end
end 

res = Collatz.sequence(5)
IO.inspect(res)

File: WINGS_AI_Bootcamp/Python/collatz1.py
''' Collatz senquence
   160,80,40,20,10,5,16,8,4,2,1,4,2,1.....
   
'''

def next_collatz(k: int) -> int:
	return k // 2 if k%2 == 0 else 3*k + 1
def collatz(n: int) -> list[int]:
	if n == 4:
		return [4,2,1]
	else:
		return [n] + collatz(next_collatz(n))
print(collatz(7))
